### PR TITLE
Restructured dev and test code

### PR DESCRIPTION
**Pull request by PiyushGSlab**
1.  Restructured the unit testing code
2.  Pulled out common code into one function in infotype_helper.py

### CODE REVIEW

File: datahub-classify/src/datahub_classify/infotype_helper.py [hunk 1]
```diff
@@ -40,32 +41,11 @@
     nlp_english = spacy.load(spacy_model_name)
 spacy_models_list = [nlp_english]
 
-# TODO: pull out the common/repetitive code
 
-
-def inspect_for_email_address(metadata, values, config):
+def compute_name_description_dtype_score(
+    metadata: Metadata, config: dict, debug_info: dict
+) -> dict:
```
Comment (hsheth2) [inline datahub-classify/src/datahub_classify/infotype_helper.py:47]: it'd be good to start using dataclasses or TypedDict instead of raw `dict` types here and elsewhere in the code

File: datahub-classify/tests/test_infotype_predictor.py [hunk 2]
```diff
@@ -42,157 +43,22 @@
...
-def get_public_data(input_data_path):
-    print("===============%s=================" % input_data_path)
-    data1 = pd.read_csv(os.path.join(input_data_path, "UCI_Credit_Card.csv"))
-    data2 = pd.read_csv(os.path.join(input_data_path, "Age2_address1_credit_card3.csv"))
-    data3 = pd.read_csv(os.path.join(input_data_path, "list_of_real_usa_addresses.csv"))
-    data4 = pd.read_csv(os.path.join(input_data_path, "CardBase.csv"))
-    data5 = pd.read_csv(os.path.join(input_data_path, "Credit_Card2.csv"))
-    data6 = pd.read_csv(os.path.join(input_data_path, "catalog.csv"))
-    data7 = pd.read_csv(os.path.join(input_data_path, "iban.csv"))
-    # data8 = pd.read_csv(os.path.join(input_data_path, "USA_cars_datasets.csv"))
```
Comment (hsheth2) [inline datahub-classify/tests/test_infotype_predictor.py:54]: why did we remove a bunch of csv files from the tests?

**Review by hsheth2 — CHANGES_REQUESTED** [review]

File: datahub-classify/tests/test_infotype_predictor.py (hunk 2 above)
Comment (PiyushGSlab) [reply datahub-classify/tests/test_infotype_predictor.py:54]: We are still using all the files, but reading them from the directory instead of referring to each file in the python code

Comment (hsheth2) [reply datahub-classify/tests/test_infotype_predictor.py:54]: It seems like a number of files were also removed from the tests directory?

File: datahub-classify/src/datahub_classify/infotype_helper.py (hunk 1 above)
Comment (hsheth2) [reply datahub-classify/src/datahub_classify/infotype_helper.py]: Not quite what I was looking for. Where possible, please use dataclasses. For example, debug_info shouldn't be a dictionary but instead an instance of a dataclass

File: datahub-classify/tests/test_infotype_predictor.py (hunk 2 above)
Comment (PiyushGSlab) [reply datahub-classify/tests/test_infotype_predictor.py:54]: Sorry for the confusion, we have removed some of the datasets which didn't have a public license (as per discussion with Swaroop)

Comment (hsheth2) [reply datahub-classify/tests/test_infotype_predictor.py:54]: ok sounds good

File: datahub-classify/src/datahub_classify/infotype_helper.py (hunk 1 above)
Comment (PiyushGSlab) [reply datahub-classify/src/datahub_classify/infotype_helper.py]: We have changed debug_info to be of TypedDict. Does this work?

**Comment by PiyushGSlab** [comment]
As per Mayuri's request, we have modified some logger statements in infotype_predictor.py, to reduce the verbosity of messages

File: datahub-classify/src/datahub_classify/infotype_helper.py [hunk 3]
```diff
@@ -152,10 +154,10 @@ def inspect_for_street_address(
 
 
 def inspect_for_gender(
-    metadata: Metadata, values: list, config: dict
-) -> tuple:  # noqa: C901
+    metadata: Metadata, values: List[Any], config: Dict[str, Any]
+) -> Tuple[float, Any]:  # noqa: C901
```
Comment (hsheth2) [inline datahub-classify/src/datahub_classify/infotype_helper.py:158]:
```suggestion
) -> Tuple[float, DebugInfo]:  # noqa: C901
```

File: datahub-classify/src/datahub_classify/infotype_helper.py [hunk 4]
```diff
@@ -181,7 +183,7 @@ def inspect_for_gender(
     try:
         if (
             debug_info.get(NAME, None)
-            and int(debug_info[NAME]) == 1
+            and int(debug_info[NAME]) == 1  # type: ignore
```
Comment (hsheth2) [inline datahub-classify/src/datahub_classify/infotype_helper.py:186]: why was this required?

File: datahub-classify/src/datahub_classify/infotype_helper.py [hunk 5]
```diff
@@ -198,10 +200,10 @@ def inspect_for_gender(
 
 
 def inspect_for_credit_debit_card_number(
-    metadata: Metadata, values: list, config: dict
-) -> tuple:
+    metadata: Metadata, values: List[Any], config: Dict[str, Any]
+) -> Tuple[float, Any]:
```
Comment (hsheth2) [inline datahub-classify/src/datahub_classify/infotype_helper.py:204]: use debuginfo

File: datahub-classify/src/datahub_classify/infotype_helper.py [hunk 6]
```diff
@@ -198,10 +200,10 @@ def inspect_for_gender(
 
 
 def inspect_for_credit_debit_card_number(
-    metadata: Metadata, values: list, config: dict
-) -> tuple:
+    metadata: Metadata, values: List[Any], config: Dict[str, Any]
```
Comment (hsheth2) [inline datahub-classify/src/datahub_classify/infotype_helper.py:203]: what is the type of config here?

**Review by hsheth2 — CHANGES_REQUESTED** [review]
Looking at the code now, a dataclass would definitely be preferable over the TypedDict

File: datahub-classify/src/datahub_classify/infotype_helper.py (hunk 6 above)
Comment (PiyushGSlab) [reply datahub-classify/src/datahub_classify/infotype_helper.py:203]: config is a type of dict (configuration for that particular infotype) 
For eg: 
 _{
        'Prediction_Factors_and_Weights': {
            'Name': 0.4,
            'Description': 0,
            'Datatype': 0,
            'Values': 0.6
        },
        'Name': { 'regex': [<regex patterns>] },
        'Description': { 'regex': [<regex patterns>] },
        'Datatype': { 'type': [<list of datatypes>] },
        'Values': {
            'prediction_type': 'regex/library',
            'regex': [<regex patterns>],
            'library': [<library name>]
        }
    },_

Comment (mardikark-gslab) [reply datahub-classify/src/datahub_classify/infotype_helper.py:203]: Hi @hsheth2 , I had a discussion with @mayurinehate on changing this dict to dataclass, in the future PR we will work on it.

File: datahub-classify/src/datahub_classify/infotype_helper.py (hunk 3 above)
Comment (PiyushGSlab) [reply datahub-classify/src/datahub_classify/infotype_helper.py]: Changed DebugInfo to dataclass

File: datahub-classify/src/datahub_classify/infotype_helper.py (hunk 4 above)
Comment (PiyushGSlab) [reply datahub-classify/src/datahub_classify/infotype_helper.py]: removed this, not required anymore

File: datahub-classify/src/datahub_classify/infotype_helper.py (hunk 5 above)
Comment (PiyushGSlab) [reply datahub-classify/src/datahub_classify/infotype_helper.py]: fixed this wherever applicable

File: datahub-classify/src/datahub_classify/infotype_helper.py [hunk 7]
```diff
@@ -339,18 +342,18 @@ def inspect_for_full_name(
...
 
     debug_info = compute_name_description_dtype_score(metadata, config, debug_info)
 
     try:
         if (
-            debug_info.get(NAME, None)
-            and int(debug_info[NAME]) == 1  # type: ignore
-            and VALUES in debug_info.keys()
-            and 0.5 > cast(float, debug_info[VALUES]) > 0.1
+            hasattr(debug_info, "name")
```
Comment (hsheth2) [inline datahub-classify/src/datahub_classify/infotype_helper.py:351]: I don't think this hasattr check is necessary anymore, how that we have the dataclass?

**Review by hsheth2 — COMMENTED** [review]
mostly looking good, had one tiny comment remaining

File: datahub-classify/src/datahub_classify/infotype_helper.py (hunk 7 above)
Comment (PiyushGSlab) [reply datahub-classify/src/datahub_classify/infotype_helper.py]: Removed the hasattr check, and also modified the condition to apply check on prediction_factors_weights which adds more clarity to the code.

**Review by hsheth2 — APPROVED** [review]
Approving for now

Note that I'm still not super happy with where we are on this codebase.
1. config parsing and handling - generally we should be using pydantic models instead of raw dicts and constants for indexes
2. class and variable naming / comments - for example, it's very unclear what DebugInfo's description actually contains (does it contain the confidence or something else?), especially since it can be a string or a float. another example: reference_input.py's input1 is actually the default config, but that fact is totally unclear from the code
3. code redundancy / duplication in the infotype_helper.py file - the inspect_* methods all still look extremely similar and should be refactored more. Ideally we should have a single large map of type -> function with the "library" implementation, and the regex and other handling should be common

**Comment by mardikark-gslab** [comment]
> Approving for now
> 
> Note that I'm still not super happy with where we are on this codebase.
> 
> 1. config parsing and handling - generally we should be using pydantic models instead of raw dicts and constants for indexes
> 2. class and variable naming / comments - for example, it's very unclear what DebugInfo's description actually contains (does it contain the confidence or something else?), especially since it can be a string or a float. another example: reference_input.py's input1 is actually the default config, but that fact is totally unclear from the code
> 3. code redundancy / duplication in the infotype_helper.py file - the inspect_* methods all still look extremely similar and should be refactored more. Ideally we should have a single large map of type -> function with the "library" implementation, and the regex and other handling should be common
> 4. The tests shouldn't have separate UNIT_TESTING and normal json files with expected values - instead we should just have a parameterized test that uses pytest marks to conditionally skip non-unit test cases

Thanks @hsheth2 for the suggestions, I have noted it. We will take care of it in the next PR.

File: datahub-classify/src/datahub_classify/infotype_helper.py [hunk 8]
```diff
@@ -185,10 +185,10 @@ def inspect_for_gender(
 
     try:
         if (
-            hasattr(debug_info, "name")
-            and int(debug_info.name) == 1
-            and hasattr(debug_info, "values")
-            and debug_info.values == 0
+            prediction_factors_weights.get(NAME, 0) > 0
+            and debug_info.name == 1.0
```
[the remark this answers is not in the document]
Comment (mardikark-gslab) [reply datahub-classify/src/datahub_classify/infotype_helper.py]: Modified the code not to use "==" for float comparison. Also I have modified DebugInfo instance variable type to float instead of float and str. Added couple of TODOs regarding adding warning/error messages  in the error flag which will get passed in ColumnInfo object in future PR.

File: datahub-classify/src/datahub_classify/infotype_helper.py [hunk 9]
```diff
@@ -186,7 +192,7 @@ def inspect_for_gender(
     try:
         if (
             prediction_factors_weights.get(NAME, 0) > 0
-            and debug_info.name == 1.0
+            and abs(1 - cast(float, debug_info.name)) < 1e-10
```
Comment (hsheth2) [inline datahub-classify/src/datahub_classify/infotype_helper.py:195]: should be using debug_info.name is not None to avoid the call to cast

Comment (mardikark-gslab) [reply datahub-classify/src/datahub_classify/infotype_helper.py]: Modified the code

File: datahub-classify/src/datahub_classify/infotype_helper.py [hunk 10]
```diff
@@ -192,7 +192,8 @@ def inspect_for_gender(
     try:
         if (
             prediction_factors_weights.get(NAME, 0) > 0
-            and abs(1 - cast(float, debug_info.name)) < 1e-10
+            and debug_info.name
```
Comment (hsheth2) [inline datahub-classify/src/datahub_classify/infotype_helper.py:195]:
```suggestion
            and debug_info.name is not None
```

File: datahub-classify/src/datahub_classify/infotype_helper.py [hunk 11]
```diff
@@ -355,7 +356,8 @@ def inspect_for_full_name(
...
```
Comment (hsheth2) [inline datahub-classify/src/datahub_classify/infotype_helper.py:359]:
```suggestion
            and debug_info.name is not None
```

**Review by hsheth2 — APPROVED** [review]
